File: backend/app/api/v1/research.py
"""Version-bound user actions for the research workbench."""
from typing import Literal

from fastapi import APIRouter, Depends, Query
from fastapi.responses import PlainTextResponse
from pydantic import BaseModel, Field, model_validator

from backend.app.dependencies import get_run_service
from deepresearch_agent.research.schemas import ResearchSpec


router = APIRouter(prefix='/research', tags=['research'])


class VersionAction(BaseModel):
    revision: int = Field(ge=1)
    fingerprint: str = Field(min_length=64, max_length=64)


class RevisionAction(VersionAction):
    spec: ResearchSpec | None = None
    instruction: str | None = Field(default=None, min_length=1, max_length=10000)

    @model_validator(mode='after')
    def one_edit(self):
        if (self.spec is None) == (self.instruction is None):
            raise ValueError('请提供结构化修改或自然语言修改其中一种')
        return self


class ApprovalAction(VersionAction):
    client_request_id: str = Field(min_length=1, max_length=128)


class CellTarget(BaseModel):
    item_id: str
    field_id: str


class FollowupAction(ApprovalAction):
    cells: list[CellTarget] = Field(min_length=1)
    reason: str = Field(min_length=1, max_length=10000)


class AcceptAction(VersionAction):
    report_fingerprint: str = Field(min_length=64, max_length=64)


@router.get('/{study_id}')
async def get_study(study_id: str, service=Depends(get_run_service)):
    study = await service.research.store.get(study_id)
    run = await service.runs.get(study['run_id']) if study.get('run_id') else None
    return {**study, 'run_status': run.status if run else None}


@router.post('/{study_id}/revisions')
async def revise(study_id: str, payload: RevisionAction, service=Depends(get_run_service)):
    return await service.research.revise(study_id, payload)


@router.post('/{study_id}/approve')
async def approve(study_id: str, payload: ApprovalAction, service=Depends(get_run_service)):
    return await service.research.approve(study_id, payload)


@router.get('/{study_id}/matrix')
async def matrix(study_id: str, service=Depends(get_run_service)):
    return await service.research.store.matrix(study_id)


@router.post('/{study_id}/followups')
async def followup(study_id: str, payload: FollowupAction, service=Depends(get_run_service)):
    return await service.research.followup(study_id, payload)


@router.post('/{study_id}/accept')
async def accept(study_id: str, payload: AcceptAction, service=Depends(get_run_service)):
    return await service.research.accept(study_id, payload)


@router.get('/{study_id}/export')
async def export(study_id: str, format: Literal['markdown', 'json'] = Query('markdown'), service=Depends(get_run_service)):
    study = await service.research.store.get(study_id)
    matrix = await service.research.store.matrix(study_id)
    report = study.get('report') or {}
    if format == 'json':
        return {'study': study, 'matrix': matrix, 'report': report}
    from deepresearch_agent.research.workflow import ResearchWorkflow
    content = str(report.get('content') or ResearchWorkflow.render(study['spec'], matrix, matrix['cells']))
    if not report.get('complete'):
        content = '> 部分结果：当前研究仍有缺口，尚未完整定稿。\n\n' + content
    return PlainTextResponse(content, media_type='text/markdown',
        headers={'Content-Disposition': f'attachment; filename="research-{study_id}.md"'})

File: frontend/src/api/research.ts
import { API_BASE } from "./client";
import { ApiError } from "../types/api";
import { ResearchMatrix, ResearchMutationResult, ResearchRevision, ResearchRevisionRef, ResearchSpec, ResearchStudy } from "../types/research";

async function researchRequest<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${API_BASE}${path}`, {
    ...init,
    headers: { "Content-Type": "application/json", ...init?.headers },
  });
  if (!response.ok) {
    const payload = await response.json().catch(() => ({ error: { code: "NETWORK_ERROR", message: response.statusText } }));
    throw new ApiError(payload.error ?? { code: "HTTP_ERROR", message: response.statusText });
  }
  return response.json() as Promise<T>;
}

const post = <T>(path: string, body: unknown) => researchRequest<T>(path, { method: "POST", body: JSON.stringify(body) });

export const researchApi = {
  study: (studyId: string) => researchRequest<ResearchStudy>(`/research/${encodeURIComponent(studyId)}`),
  revisions: (studyId: string) => researchRequest<ResearchRevisionRef[]>(`/research/${encodeURIComponent(studyId)}/revisions`),
  revision: (studyId: string, revision: number) => researchRequest<ResearchRevision>(`/research/${encodeURIComponent(studyId)}/revisions/${revision}`),
  restore: (study: ResearchStudy, sourceRevision: number) => post<ResearchStudy>(`/research/${encodeURIComponent(study.study_id)}/restore`, {
    revision: study.current_revision, fingerprint: study.fingerprint, source_revision: sourceRevision,
  }),
  matrix: (studyId: string) => researchRequest<ResearchMatrix>(`/research/${encodeURIComponent(studyId)}/matrix`),
  reviseSpec: (study: ResearchStudy, spec: ResearchSpec) => post<ResearchStudy>(`/research/${encodeURIComponent(study.study_id)}/revisions`, {
    revision: study.current_revision, fingerprint: study.fingerprint, spec,
  }),
  reviseInstruction: (study: ResearchStudy, instruction: string) => post<ResearchStudy>(`/research/${encodeURIComponent(study.study_id)}/revisions`, {
    revision: study.current_revision, fingerprint: study.fingerprint, instruction,
  }),
  approve: (study: ResearchStudy) => post<ResearchMutationResult>(`/research/${encodeURIComponent(study.study_id)}/approve`, {
    revision: study.current_revision, fingerprint: study.fingerprint, client_request_id: crypto.randomUUID(),
  }),
  followup: (study: ResearchStudy, cells: Array<{item_id: string; field_id: string}>, reason: string) => post<ResearchMutationResult>(`/research/${encodeURIComponent(study.study_id)}/followups`, {
    revision: study.current_revision, fingerprint: study.fingerprint, client_request_id: crypto.randomUUID(), cells, reason,
  }),
  accept: (study: ResearchStudy) => post<ResearchStudy>(`/research/${encodeURIComponent(study.study_id)}/accept`, {
    revision: study.current_revision, fingerprint: study.fingerprint, report_fingerprint: study.report?.fingerprint,
  }),
  exportUrl: (studyId: string, format?: "json") => `${API_BASE}/research/${encodeURIComponent(studyId)}/export${format ? `?format=${format}` : ""}`,
};
